Implements Cumprod function for autograd

martinarjovsky opened the pull request:
In reference to https://github.com/pytorch/pytorch/issues/1095 :)

The algorithm for nonzero elements is straightforward and pretty fast, but when zero elements
are present is pretty hard to do vectorized (and O(n^2) with n elements in the dimension of the input where the cumprod is done), hence the code is a bit complex.

A set of extra tests passed is present here https://gist.github.com/martinarjovsky/1d4679c54b2fc5cddf73cdd45b359c9f

Commit:
Initial setup for cumprod

Commit:
[WIP] implemented backward for cumprod when input is nonzero

Commit:
Finishing implementation of cumprod, tests passed but not included

Commit:
adding tests

Commit:
Fixing final bugs, all tests passed

albanD commented on torch/autograd/_functions/tensor.py:
```diff
@@ -601,13 +609,93 @@ def forward(self, input):
+            LT = torch.LongTensor
+
+        if (input == 0).any():
+            ones = input.index_select(self.dim,
```
Why are you using `index_select` with a single element instead of `select` here? If you use select, you don't need to create a LongTensor every time.
If the reason is to keep a dimension of size 1, you can use `narrow` with a length of 1 to avoid creating all the LongTensors.

albanD commented on torch/autograd/_functions/tensor.py:
```diff
@@ -601,13 +609,93 @@ def forward(self, input):
+                # At this point ommitted_products is the same size
+                # as input, except on the dimension dim where it's
+                # dim_size - k
+                assert ommitted_products.size()[self.dim] == dim_size - k, \
```
`ommitted_products.size(self.dim)`

albanD commented on torch/autograd/_functions/tensor.py:
```diff
@@ -601,13 +609,93 @@ def forward(self, input):
+                        torch.sum(grad_output * ommitted_products, dim=self.dim))
 
+        else:
+            output = torch.cumprod(input, dim=self.dim)
```
You can save this during the forward pass so that you don't have to recompute it here.

martinarjovsky replied:
I thought about this, but the computational cost of recalculating it seems small. The added cost is doing one cumprod, when we are already making 5 or 6 other operations with the same cost, including one cumsum. Right now we pay a bit more of overhead but store less things for the backward, saving half the memory.

Thoughts?

albanD commented on torch/autograd/_functions/tensor.py:
```diff
@@ -601,13 +609,93 @@ def forward(self, input):
+        
+class Cumprod(Function):
+    
+    def __init__(self, dim):
```
This is an old style autograd function, @apaszke do we absolutely want this to be a new style class or is it ok?

apaszke commented on torch/autograd/_functions/tensor.py:
```diff
@@ -601,13 +609,93 @@ def forward(self, input):
+                        LT(range(k))), dim=self.dim)
+
+                    prods_from_k_plus_1 = torch.cumprod(input.index_select(
+                        self.dim, LT(range(k+1, dim_size))), dim=self.dim)
```
Creating tensors from Python sequences is slow and GPU unfriendly (it's a host -> device copy at every iteration). I'd use `torch.arange` (+ an optional GPU copy) before the loop, and then use `idx[:k]` instead of `LT(range(k))` and `idx[:k + 1:dim_size]` instead of `LT(range(k+1, dim_size))`.

martinarjovsky replied:
This makes a lot of sense, will do.

martinarjovsky replied:
I implemented this, only with idx = torch.LongTensor(range(k)) instead of torch.arange returns a FloatTensor and index_select requires a LongTensor. Hope it's okay!

apaszke commented on torch/autograd/_functions/tensor.py:
```diff
@@ -601,13 +609,93 @@ def forward(self, input):
+                        LT(range(k))), dim=self.dim)
+
+                    ommitted_products = prods_until_k
+                else: #k == 0
```
Wouldn't it be simpler to order these cases like this?
```python
if k == 0:
    ...
elif k == dim_size - 1:
    ...
else:
    ...
```

apaszke commented on torch/autograd/_functions/tensor.py:
```diff
@@ -601,13 +609,93 @@ def forward(self, input):
+        else:
+            LT = torch.LongTensor
+
+        if (input == 0).any():
```
Instead of having this massive `if else`, I'd do it like that:
```python
if (input != 0).all():
    # simple case
    grad_input = ...
    return grad_input

# complicated case
```

apaszke commented on torch/autograd/_functions/tensor.py:
```diff
@@ -601,13 +609,93 @@ def forward(self, input):
+                    prods_from_k_plus_1 = torch.cumprod(input.index_select(
+                        self.dim, LT(range(k+1, dim_size))), dim=self.dim)
+
+                    ommitted_products = prods_until_k.expand_as(
```
Can you add some comments or at least give a brief description of the algorithm in this PR? It would make reviewing much easier

martinarjovsky replied:
Will do!

apaszke commented on torch/autograd/_functions/tensor.py:
```diff
@@ -601,13 +609,93 @@ def forward(self, input):
+                # as input, except on the dimension dim where it's
+                # dim_size - k
+                assert ommitted_products.size()[self.dim] == dim_size - k, \
+                    "Dimension error"
```
also, no need for that message (the condition tells much more than the msg)

apaszke commented on torch/autograd/_functions/tensor.py:
```diff
@@ -601,13 +609,93 @@ def forward(self, input):
+                    "Dimension error"
+
+                if k != 0:
+                    size_to_expand = [l for l in input.size()]
```
```python
size_to_expand = list(input.size())
```

apaszke commented on torch/autograd/_functions/tensor.py:
```diff
@@ -601,13 +609,93 @@ def forward(self, input):
+                    size_to_expand = [l for l in input.size()]
+                    size_to_expand[self.dim] = k # Adding zeros to missing dimensions
+                    size_to_expand = torch.Size(size_to_expand)
+                    expanded_zeros = zeros.expand(size_to_expand)
```
this seems to be unused

fmassa commented:
Did we consider using the implementation of `cumsum` to compute `cumprod`? We could writesomething like
```python
def cumprod(x):
    return exp(cumsum(log(x)))
```
Maybe it's very unstable computing it like that?

apaszke commented:
@fmassa I'd expect it to be very very unstable for `x` close to 0 (+ cumprod is well defined when some values are 0, but your definition doesn't handle that case - log will give you `-inf`s)

fmassa commented:
@apaszke I'm not so sure that it would be very very unstable for x close to 0, the forward pass doesn't seem to be that sensitive.
My implementation doesn't handle the cases where `x <= 0`, but that can be easily fixed I think.
Anyway, those are my 2 cents on the matter. :)

martinarjovsky commented on torch/autograd/_functions/tensor.py:
```diff
@@ -601,13 +609,93 @@ def forward(self, input):
+
+        if (input == 0).any():
+            ones = input.index_select(self.dim,
+                    LT(range(1))).fill_(1).clone()
```
Actually I think there's a bug in here. The clone should be before the fill, otherwise I'm now modifying input to have a bunch of ones.

apaszke replied:
You're not - `index_select` returns a new tensor that doesn't share memory with input.

martinarjovsky replied:
Ah you're right, my bad. The clone is necessary though when using select rather than index_select right?

apaszke replied:
Correct

martinarjovsky commented:
@fmassa This would have two problems.
- It would support negative inputs.
- It wouldn't support zero inputs. The case with zero inputs is handled in two lines already by doing
`
output = torch.cumprod(input, dim=self.dim)
return reverse_cumsum(output * grad_output, dim=self.dim) / input` 

It's the case with zeros that's hard :)

Commit:
Fixing linting issues + the ones in the review

martinarjovsky commented:
Just fixed (or at least attempted) the issues. I added a detailed comment of the algorithm for the backward pass but it's likely way to much. Should I take it out and put it in the PR instead?

Commit:
adding clone when returning grad_output

apaszke commented on torch/autograd/_functions/tensor.py:
```diff
@@ -601,13 +611,168 @@ def forward(self, input):
+            idx = torch.LongTensor(range(dim_size))
+
+        ones = input.select(self.dim, 0).unsqueeze(self.dim).clone().fill_(1)
+        zeros = ones.clone().fill_(0)
```
These two are more efficient (right now it's copy + fill, but the copy is unnecessary):
* `ones.new().resize_as_(ones).fill_(0)`
* `ones * 0`

martinarjovsky replied:
Done. I had to create a size variable though (for ones, not zeros), which is a bit ugly. Any way to avoid this?

apaszke commented on torch/autograd/_functions/tensor.py:
```diff
@@ -591,6 +591,16 @@ def backward(self, grad_output):
+    ret = torch.cumsum(-x, dim=dim)
+
+    end_idx = ret.size(dim) - 1
+    ret_sum = ret.narrow(dim, end_idx, 1)
```
Can you add a `clone()` here? I just realized that `ret` and `ret_sum` share storage, so `-=` doesn't necessarily has to be valid (apparently it is at the moment, but we don't guarantee it).

martinarjovsky replied:
As a small comment, this will also affect the current behaviour of Cumsum when merged (the clone is missing in master for Cumsum as well).

apaszke commented on torch/autograd/_functions/tensor.py:
```diff
@@ -591,6 +591,16 @@ def backward(self, grad_output):
         return grad_input
 
 
+def reverse_cumsum(x, dim):
```
Can you please rename it to `sum_scan_exclusive`?

martinarjovsky replied:
Done! Why the name though? :P

apaszke replied:
https://en.wikipedia.org/wiki/Prefix_sum - cumsum is an inclusive scan using +, this function is an exclusive scan using +.

martinarjovsky replied:
Gotcha

apaszke commented on torch/autograd/_functions/tensor.py:
```diff
@@ -601,13 +611,168 @@ def forward(self, input):
+                    self.dim
+                )
+
+            grad_input.index_copy_(
```
I think you could do it like that (regular `copy_` is likely to be faster than `index_copy_`):
```python
grad_input.select(k, self.dim).copy_(torch.sum(...))
```

apaszke commented on torch/autograd/_functions/tensor.py:
```diff
@@ -601,13 +611,168 @@ def forward(self, input):
+        else:
+            idx = torch.LongTensor(range(dim_size))
+
+        ones = input.select(self.dim, 0).unsqueeze(self.dim).clone().fill_(1)
```
Same as below, it'd be best to try to avoid the copy that `clone` incurs.

apaszke commented on torch/autograd/_functions/tensor.py:
```diff
@@ -601,13 +611,168 @@ def forward(self, input):
+        if input.is_cuda:
+            idx = torch.cuda.LongTensor(range(dim_size))
+        else:
+            idx = torch.LongTensor(range(dim_size))
```
Actually I think creating these tensors could be avoided (it's best to not initialize them with Python objects -- it's **slow**). The only places where you use them is `index_select`, but you only slice them, so a faster and equivalent way would be to do this:
```python
dim_padding = (slice(None, None),) * (self.dim - 1)
# now you can do this instead of input.index_select(self.dim, idx[k + 1:dim_size])
input[dim_padding + (slice(k+1, None),)]
```

This doesn't copy the memory and avoids the `index_select` kernel which is not a trivial operation.

martinarjovsky replied:
Done! Good call.

apaszke commented on torch/autograd/_functions/tensor.py:
```diff
@@ -601,13 +611,168 @@ def forward(self, input):
+                ommitted_products = torch.cat(
+                    (zeros.expand(size_to_expand), ommitted_products),
+                    self.dim
+                )
```
Maybe it would be better to just slice the grad_output, Instead of adding zero padding and doing an element-wise mul

martinarjovsky replied:
Done, seems reasonable.

Commit:
Adding efficiency and style changes

martinarjovsky committed:
Merge branch 'master' into master

martinarjovsky committed:
Fixing linting issue

colesbury commented:
@pytorchbot add to whitelist

martinarjovsky commented:
@apaszke Anything else I should do? :)

apaszke commented:
Sorry, we're all busy with v0.2 features. I'll try to review it soon and let you know, but it should be good.

apaszke commented on torch/autograd/_functions/tensor.py:
```diff
@@ -565,12 +575,156 @@ def forward(self, input):
+        input, = self.saved_tensors
+        dim_size = input.size(self.dim)
+        if dim_size == 1:
+            return grad_output.clone()
```
no need to clone it here

apaszke commented on torch/autograd/_functions/tensor.py:
```diff
@@ -565,12 +575,156 @@ def forward(self, input):
+
+        ones_size = list(input.size())
+        ones_size[self.dim] = 1
+        ones = input.new().resize_(ones_size).fill_(1)
```
I just realized that this can be a single 1 that is expanded to `ones_size`. No need to allocate such a large tensor

martinarjovsky replied:
I didn't quite get this. Could you elaborate a bit? I need to concatenate the 'ones' tensor later.

martinarjovsky replied:
Done :)

apaszke commented on torch/autograd/_functions/tensor.py:
```diff
@@ -565,12 +575,156 @@ def forward(self, input):
+        ones_size = list(input.size())
+        ones_size[self.dim] = 1
+        ones = input.new().resize_(ones_size).fill_(1)
+        zeros = ones * 0
```
this can also be a single zero that is expanded

martinarjovsky replied:
Actually after one of your earlier fixes the 'zeros' tensor is no longer used :P

apaszke commented on torch/autograd/_functions/tensor.py:
```diff
@@ -565,12 +575,156 @@ def forward(self, input):
+
+            grad_input.select(self.dim, k).copy_(torch.sum(
+                grad_output[dim_padding + (slice(k, None),)] * omitted_products,
+                dim=self.dim).squeeze())
```
Can you remove this `squeeze`? It's going to be unnecessary after the broadcasting PR

martinarjovsky committed:
Merge branch 'master' into master

Commit:
[WIP] Finished merging with master

Commit:
Removing extra zero tensor

Commit:
Finished, adding expand for ones

apaszke commented:
Thank you!
